File: java/robot-name/src/main/java/Robot.java
import java.util.HashSet;
import java.util.Random;
import java.util.Set;

/**
 * @author Cedric Zhuang
 */
public class Robot {
    private static final Random randGen = new Random();
    private static final Set<String> usedNames = new HashSet<>();
    private String name = "";

    Robot() {
        reset();
    }

    void reset() {
        char cName[] = new char[5];
        while (name.isEmpty() || usedNames.contains(name)) {
            cName[0] = rAlpha();
            cName[1] = rAlpha();
            cName[2] = rDigit();
            cName[3] = rDigit();
            cName[4] = rDigit();
            name = String.valueOf(cName);
        }
        usedNames.add(name);
    }

    String getName() {
        return name;
    }

    private char rChar(char start, int range) {
        return (char) ((int) start + randGen.nextInt(range));
    }

    private char rAlpha() {
        return rChar('A', 26);
    }

    private char rDigit() {
        return rChar('0', 10);
    }
}

File: scala/robot-name/src/main/scala/Robot.scala
import scala.annotation.tailrec
import scala.util.Random

/**
  *
  * @author Cedric Zhuang
  *
  */
class Robot {
  private val chars: Array[Char] = ('A' to 'Z').toArray
  private val numbers: Array[Char] = ('0' to '9').toArray

  var name: String = nameGen

  def reset(): Unit = name = nameGen

  private def randChar: Char = chars(Random.nextInt(chars.length))

  private def randInt: Char = numbers(Random.nextInt(numbers.length))

  @tailrec
  private def nameGen: String = {
    val ret: String = s"$randChar$randChar$randInt$randInt$randInt"
    if (Robot.usedNames.contains(ret))
      nameGen
    else {
      Robot.usedNames += ret
      ret
    }
  }
}

object Robot {
  private var usedNames: Set[String] = Set.empty
}
